src/consensus/blockchain.py

```diff
@@ -171,7 +176,7 @@ def contains_sub_block(self, header_hash: bytes32) -> bool:
         True if we have already added this block to the chain. This may return false for orphan sub-blocks
         that we have added but no longer keep in memory.
         """
-        return header_hash in self.sub_blocks
```

Comment:
This is wrong, this method should return true if the header hash in the blockchain. I suggest having a "sub height to hash" equivalent data structure which just conntains the header hashes, for quick lookup